File: Java Galaga/Enemy.java
import java.awt.*;//must be imported to use Graphics and Color
import java.util.ArrayList;

import javax.swing.*;

public class Enemy extends DrawableObject
{
    private int vx;                   // x velocity
    private int vy;
    private int numMoves;
    private int maxMoves;
    private int width;
    private int height;

    protected boolean allowedToShoot = false;
    public Enemy(Image img, int xLoc, int yLoc )
    {
        x = xLoc;
        y = yLoc;
        vx = 1;
        vy = 20;
        numMoves = 0;
        maxMoves = 150;
        image = img;
        width = image.getWidth(null);
        height = image.getHeight(null);
        
        
    }

    public Enemy(String path, int xLoc, int yLoc)
    {
        this(new ImageIcon(path).getImage(), xLoc, yLoc);
        vx = 1;
        vy = 20;
        numMoves = 0;
        maxMoves = 150;
    }

    public Projectile update(boolean turnToShoot, boolean enCanShoot, boolean gridMovingRight, ArrayList<Projectile> playerBullets, Game game)
    {

        for (Projectile bullet: playerBullets)
        {
            if (bullet.isInside(this)) {
                return bullet;
            }
        }
        move();
        if (turnToShoot && enCanShoot && allowedToShoot) {
            game.enemyShoot(x + 10, y);
        }
        
        return null;
    }

    public void move()
    {
        x += vx;
        if(numMoves > maxMoves)
        {
            vx = -vx;
            y += vy;
            numMoves = 0;
            maxMoves = 300;
        } 
        numMoves++;
    }

    @Override
    public void draw( Graphics page )
    {
        page.drawImage(image, x - 5, y, null);
    }
    
    public int getWidth()
    {
    	return width;
    	
    }
    
    public int getHeight()
    {
    	return height;
    }
}

File: python/galaga_gym/envs/galaga_env/DeepLearningEnv/Enemy.py
from Object import Object

class Enemy(Object):
  def __init__(self, img, x=0, y=0):
    super().__init__(img, x, y)
    self.vx = 1
    self.vy = 20
    numMoves = 0
    maxMoves = 150
    self.allowedToShoot = False
  
  def update(self, turnToShoot, enCanShoot, gridMovingRight, playerBullets, game):
    for bullet in playerBullets:
      if bullet.inside(self):
        return bullet
    
    self.move()

    if turnToShoot and enCanShoot and allowedToShoot:
      game.enemyShoot(self.x + 10, self.y)
    
    return None
  
  def move(self):
    self.x += self.vx

    if self.numMoves > self.maxMoves:
      self.vx = -self.vx
      self.y += self.vy
      self.numMoves = 0
      self.maxMoves = 300
    
    self.numMoves += 1
  
  def draw(self, screen):
    screen.blit(self.img, (self.x - 5, self.y, self.width, self.height))